src/paths/steiner.cpp

```diff
@@ -0,0 +1,211 @@
+#include "igraph_paths.h"
+#include "igraph_adjlist.h"
+#include "igraph_types.h"
+#include "igraph_memory.h"
+#include "igraph_error.h"
+#include "igraph.h"
+
+#include <string.h>
+#include <math.h>
+#include <map>
+#include <limits.h>
+#include <vector>
+#include <set>
+
+std::map<std::set<igraph_integer_t>, igraph_integer_t> subsetMap;
+
+std::vector<std::set<igraph_integer_t>> generateSubsets(igraph_vector_t steinerTerminals, igraph_integer_t n, igraph_integer_t graphsize)
+{
+	igraph_integer_t count = pow(2, n);
+	std::vector<std::set<igraph_integer_t>> allSubsets;
+	igraph_integer_t subsetIndex = graphsize;
+
+	// The outer for loop will run 2^n times to print all subset .
+	// Here variable i will act as a binary counter
+
+	for (igraph_integer_t i = 0; i < count; i++)
+	{
+		// The inner for loop will run n times , As the maximum number of elements a set can have is n
+		// This loop will generate a subset
+		std::set<igraph_integer_t> newSubset;
+		for (igraph_integer_t j = 0; j < n; j++)
+		{
+			// This if condition will check if jth bit in binary representation of  i  is set or not
+			// if the value of (i & (1 << j)) is greater than 0 , include arr[j] in the current subset
+			// otherwise exclude arr[j]
+			if ((i & (1 << j)) > 0)
+			{
+				newSubset.insert(VECTOR(steinerTerminals)[j]);
+			}
+		}
+
+		if (newSubset.size() > 1)
+		{
+			allSubsets.push_back(newSubset);
+			subsetMap.insert(std::make_pair(newSubset, subsetIndex));
+			subsetIndex++;
+		}
+	}
+
+	return allSubsets;
+}
+
+igraph_integer_t fetchIndexofMapofSets(std::set<igraph_integer_t> subset)
+{
+	igraph_integer_t key;
+	std::map<std::set<igraph_integer_t>, igraph_integer_t>::iterator it;
+	for (it = subsetMap.begin(); it != subsetMap.end(); ++it)
+	{
+		if (it->first == subset)
+		{
+			key = it->second;
+		}
+	}
+
+	return key;
+}
+
+igraph_error_t igraph_steiner_dreyfus_wagner(const igraph_t *graph,igraph_vector_t steiner_terminals,
+igraph_neimode_t mode, const igraph_vector_t *weights)
+{
+
+	igraph_integer_t no_of_vertices = (igraph_integer_t)igraph_vcount(graph);
+	igraph_integer_t no_of_edges = igraph_ecount(graph);
+
+	igraph_vector_t steiner_vertices;
+	igraph_matrix_t dp_cache; // dynamic programming table
+	igraph_integer_t q;
+	std::vector<std::set<igraph_integer_t>> allSubsets;
+	igraph_matrix_t distance;
+
+	IGRAPH_VECTOR_INIT_FINALLY(&steiner_vertices, 0);
+
+	if (igraph_vector_size(weights) != no_of_edges)
+	{
+		IGRAPH_ERROR("Weight vector length does not match", IGRAPH_EINVAL);
+	}
+
+	igraph_shortest_paths_johnson(graph, &distance, igraph_vss_all(), igraph_vss_all(), weights);
```

Comment:
With all igraph functions, except constructors, any data structure that is passed in must be initialized. Here, `distance` not initialized, hence the segmentation fault.